### PR TITLE
Load and coerce data using schema

### PR DESCRIPTION
**Pull request by Eleonore9**
**Aim**: I need to be able to load data with the right data type in each column, hence data coercion with schema

**Issue**: I takes a long time to run and fail -> issue #9

**Investigation**:
* Researched and schema validation seems to block -> Improve schema coercion
* Coercion faster, but dataset creation fails `ExceptionInfo Can't create dataset from incomplete maps` -> It must me due to missing data in a few rows -> addresses issue #5 

**Solution**:
* Improved schema validation
* Added option to handle errors due to inconsistent data

*Notes*: 
* Running `csv-to-dataset` on the repair data is still long
* The relevance of the second arity in `csv-to-dataset` can be discuss

### CODE REVIEW

File: src/kixi/lapita.clj [hunk 1]
```diff
@@ -20,38 +20,57 @@
...
+  ([filename]
+   (ds/dataset (csv-to-maps-coll filename)))
+  ;; Returns a mix of coerced data and data inconsistent w/ schema
+  ([filename schema]
+   (let [all-data (csv-to-maps-coll filename)
+         coerced-data (sc/coerce-data-from-schema all-data schema)]
+     (ds/dataset coerced-data)))
+  ;; Specify options in the 3rd arg to print number of rows with inconsistent data,
+  ;; write inconsistent data to file return only coerced data.
+  ([filename schema {:keys [print-errors write-errors remove-errors]}]
```
Comment (acron0) [inline src/kixi/lapita.clj]: When you have an arity with optional additional 'options' it's usual that the version without (two-arity in this case) calls the version with (three-arity in this case), just with an empty options map:

```
(defn foo
  ([x y opts]
    ...)
  ([x y]
    (foo x y {}))
```

File: src/kixi/lapita.clj [hunk 2]
```diff
@@ -20,38 +20,57 @@
...
+   (let [all-data (csv-to-maps-coll filename)
+         coerced-data (sc/coerce-data-from-schema all-data schema)
+         errors (when (or print-errors write-errors) (sc/gather-errors coerced-data))]
+     (when print-errors (println (format "There are %d rows with data coercion errors out of %d rows"
+                                         (count errors) (count coerced-data))))
+     (when write-errors
+       (write-csv! errors "data/records-with-data-coercion-issues.csv")
```
Comment (acron0) [inline src/kixi/lapita.clj]: Either save in a temp (`/tmp') file or allow the user to specify a location.

File: src/kixi/lapita.clj [hunk 3]
```diff
@@ -109,13 +128,24 @@
        ds/dataset))
 
 ;; Examples of uses
-(comment (def repairs-data (load-csv "data/historic-repairs-2014-2015.csv"))
+(comment (def repairs-data (csv-to-maps-coll "data/historic-repairs-2014-2015.csv"))
```
Comment (acron0) [inline src/kixi/lapita.clj]: Do these still need to be here? Are they in tests now?

**Review by acron0 — CHANGES_REQUESTED** [review]

File: src/kixi/lapita.clj (hunk 1 above)
Comment (Eleonore9) [reply src/kixi/lapita.clj]: Oh yes, that's true! Thanks!

File: src/kixi/lapita.clj (hunk 3 above)
Comment (Eleonore9) [reply src/kixi/lapita.clj]: Well, they're very handing for me. But I guess I should save those examples locally!

File: src/kixi/lapita.clj [hunk 4]
```diff
@@ -53,19 +53,18 @@
...
   ([filename schema]
-   (let [all-data (csv-to-maps-coll filename)
-         coerced-data (sc/coerce-data-from-schema all-data schema)]
-     (ds/dataset coerced-data)))
+   (csv-to-dataset filename schema {}))
   ;; Specify options in the 3rd arg to print number of rows with inconsistent data,
   ;; write inconsistent data to file return only coerced data.
-  ([filename schema {:keys [print-errors write-errors remove-errors]}]
+  ([filename schema {:keys [print-errors write-errors errors-filepath remove-errors]
+                     :or {errors-filepath "test-data/records-with-data-coercion-issues.csv"}}]
```
Comment (acron0) [inline src/kixi/lapita.clj]: This is still not a safe file path to assume. You either need to force them to specify a dir or create a temporary file using a safe method.

File: src/kixi/lapita.clj [hunk 5]
```diff
@@ -53,19 +53,18 @@
...
    (let [all-data (csv-to-maps-coll filename)
          coerced-data (sc/coerce-data-from-schema all-data schema)
          errors (when (or print-errors write-errors) (sc/gather-errors coerced-data))]
      (when print-errors (println (format "There are %d rows with data coercion errors out of %d rows"
                                          (count errors) (count coerced-data))))
      (when write-errors
-       (write-csv! errors "data/records-with-data-coercion-issues.csv")
+       (write-csv! errors errors-filepath)
        (println "The records with data coercion issues were saved in 'data/records-with-data-coercion-issues.csv'"))
```
Comment (acron0) [inline src/kixi/lapita.clj]: Output should use `errors-filepath`

**Review by acron0 — CHANGES_REQUESTED** [review]

File: src/kixi/lapita/schemas.clj [hunk 6]
```diff
@@ -1,39 +1,49 @@
...
+  "Performs the data coercion using a schema.
+   If the data can be coerced, it's returned w/ :validated metadata associated.
+   If there's a coercion error the non coerced data is returned associated w/ :errors
+   metadata."
+  [data schema]
+  (let [data-coercer (coerce/coercer schema coerce/string-coercion-matcher)]
+    (pmap (fn [d] (let [coerced (data-coercer d)]
+                    (if (su/error? coerced)
+                      (with-meta d {:data-coercion :errors})
+                      (with-meta coerced {:data-coercion :validated}))))
```
Comment (acron0) [inline src/kixi/lapita/schemas.clj]: Could we avoid using metadata here and just return a map with `:results` and `:errors`?

File: src/kixi/lapita.clj [hunk 7]
```diff
@@ -111,7 +124,7 @@
 (comment (def repairs-data (load-csv "data/historic-repairs-2014-2015.csv"))
```
Comment (acron0) [inline src/kixi/lapita.clj]: Remove these pls

File: src/kixi/lapita/schemas.clj [hunk 8]
```diff
@@ -1,13 +1,18 @@
...
-            [schema-contrib.core :as c]))
+            [schema-contrib.core :as c]
+            [schema.utils :as su]))
 
 ;; Create and valid schemas using Plumatic schemas (https://github.com/plumatic/schema)
 (defn coerce-data-from-schema
   [data schema]
   (let [data-coercer (coerce/coercer schema coerce/string-coercion-matcher)]
-    (pmap data-coercer data)))
+    (pmap (fn [d] (let [coerced (data-coercer d)]
```
Comment (acron0) [inline src/kixi/lapita/schemas.clj]: I don't think you need this pmap at all now

File: src/kixi/lapita/schemas.clj [hunk 9]
```diff
@@ -1,39 +1,52 @@
...
 
-(defn make-col-names-schema
-  [col-schema]
-  (mapv (fn [s] (let [datatype (:schema s)
-                      fieldname (:name s)]
-                  (s/one datatype fieldname)))
-        (:column-names col-schema)))
+;; NOTE: schema-contrib.core/Date expect a date in format ISO 8601 like YYYY-MM-DD
+;; To solve that, change the format on the spreadsheet to match ISO 8601.
+(def HousingRepairSchema
```
Comment (acron0) [inline src/kixi/lapita/schemas.clj]: Move to tests

File: src/kixi/lapita.clj [hunk 10]
```diff
@@ -20,38 +20,56 @@
...
+   that fails data coercion.
+   Example of options {:print-errors true :write-errors false :remove-errors true}"
+  ([filename]
+   (ds/dataset (csv-to-maps-coll filename)))
+  ;; Returns a mix of coerced data and data inconsistent w/ schema
+  ([filename schema]
+   (csv-to-dataset filename schema {}))
+  ;; Specify options in the 3rd arg to print number of rows with inconsistent data,
+  ;; write inconsistent data to file return only coerced data.
+  ([filename schema {:keys [print-errors write-errors errors-filepath remove-errors]}]
```
Comment (acron0) [inline src/kixi/lapita.clj]: I wouldn't bother having `write-errors` and `errors-filepath`, just have `write-errors` and let it either be a path to write to or nil.

**Review by acron0 — CHANGES_REQUESTED** [review]

File: src/kixi/lapita/schemas.clj (hunk 9 above)
Comment (Eleonore9) [reply src/kixi/lapita/schemas.clj]: I'm not using it in the test. Would it be a good place for my schema?

**Review by acron0 — APPROVED** [review]
looking good!

**Pull request merged**